deps/oauth2_client/src/oauth2_client.erl

```diff
@@ -0,0 +1,466 @@
+        {error,_} = Error -> Error;
+        _ -> map_to_oauth_provider(Config)
+      end,
+      rabbit_log:debug("Resolved oauth_provider ~p", [OAuthProvider]),
+      case find_missing_attributes(OAuthProvider, ListOfRequiredAttributes) of
+        [] -> {ok, OAuthProvider};
+        _ -> Result2 = case OAuthProvider#oauth_provider.issuer of
+                undefined -> {error, {missing_oauth_provider_attributes, [issuer]}};
+                Issuer ->
+                    rabbit_log:debug("Downloading oauth_provider ~p using issuer ~p", [OAuth2ProviderId, Issuer]),
+                    case get_openid_configuration(Issuer, get_ssl_options_if_any(OAuthProvider)) of
+                      {ok, OauthProvider} -> {ok, update_oauth_provider_endpoints_configuration(OAuth2ProviderId, OauthProvider)};
+                      {error, _} = Error2 -> Error2
+                    end
+            end,
+            case Result2 of
+              {ok, OAuthProvider2} ->
+                  case find_missing_attributes(OAuthProvider2, ListOfRequiredAttributes) of
+                    [] ->
+                      rabbit_log:debug("Resolved oauth_provider ~p", [OAuthProvider]),
+                      {ok, OAuthProvider2};
+                    _ = Attrs->
+                      {error, {missing_oauth_provider_attributes, Attrs}}
+                  end;
+              {error, _} = Error3 -> Error3
+            end
+      end
+  end.
+
+%% HELPER functions
+
+
+oauth_provider_to_proplists(#oauth_provider{} = OAuthProvider) ->
+  lists:zip(record_info(fields, oauth_provider), tl(tuple_to_list(OAuthProvider))).
+filter_undefined_props(PropList) ->
+  lists:foldl(fun(Prop, Acc) ->
+    case Prop of
+      {Name, undefined} -> Acc ++ [Name];
+      _ -> Acc
+    end end, [], PropList).
+
+-spec intersection(list(), list()) -> list().
+intersection(L1, L2) ->
+  S1 = sets:from_list(L1),
+  S2 = sets:from_list(L2),
+  lists:usort(sets:to_list(sets:intersection(S1, S2))).
+
+find_missing_attributes(#oauth_provider{} = OAuthProvider, RequiredAttributes) ->
+  PropList = oauth_provider_to_proplists(OAuthProvider),
+  Filtered = filter_undefined_props(PropList),
+  intersection(Filtered, RequiredAttributes).
+
+lookup_oauth_provider_from_keyconfig() ->
+  Issuer = application:get_env(rabbitmq_auth_backend_oauth2, issuer, undefined),
+  TokenEndpoint = application:get_env(rabbitmq_auth_backend_oauth2, token_endpoint, undefined),
+  Map = maps:from_list(application:get_env(rabbitmq_auth_backend_oauth2, key_config, [])),
+  #oauth_provider{
+    issuer = Issuer,
+    jwks_uri = maps:get(jwks_url, Map, undefined), %% jwks_url not uri . _url is the legacy name
+    token_endpoint = TokenEndpoint,
+    ssl_options = extract_ssl_options_as_list(Map)
+  }.
+
+-spec extract_ssl_options_as_list(#{atom() => any()}) -> proplists:proplist().
+extract_ssl_options_as_list(Map) ->
+  Verify = case maps:get(cacertfile, Map, undefined) of
+    undefined -> verify_none;
```

Comment:
Should we attempt to use https://www.erlang.org/doc/man/public_key#cacerts_get-0 to load system certificates by default if no CA certs are provided? Then we can always use `verify_none` if that function fails as well. This would be more secure by default as system certificates are likely to be available.

Reply:
You are right @lhoguin . I will make that change. Furthermore, I think we should log warning message when the user sets `peer_vertification` to `verify_peer` and we override it to `verify_none` when there are no CA certs available.

Reply:
> Should we attempt to use https://www.erlang.org/doc/man/public_key#cacerts_get-0 to load system certificates by default if no CA certs are provided

We should do that everywhere in RabbitMQ, now that `public_key:cacerts_get/0` exists ([here is one example why](https://github.com/erlang/otp/issues/8066)). I will file an issue.

Reply:
https://github.com/rabbitmq/rabbitmq-server/issues/10519